src/wasi/Wasi.cpp

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2023-present Samsung Electronics Co., Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "wasi/Wasi.h"
+
+#define own
+
+namespace Walrus {
+
+wasm_extern_t* Wasi::register_functions(std::string& funcName)
+{
+    wasiFunction* func = nullptr;
+    for (unsigned i = 0; i < m_wasiFunctions.size(); ++i) {
+        if (strcmp(m_wasiFunctions[i].name.c_str(), funcName.c_str()) == 0) {
+            func = &m_wasiFunctions[i];
+        }
+    }
+
+    if (func == nullptr) {
+        printf("Cannot find function: %s\n", funcName.c_str());
+        RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    wasm_engine_t* _engine = wasm_engine_new();
+    wasm_store_t* _store = wasm_store_new(_engine);
+
+    size_t argSize = func->funcArgTypes.size();
+    wasm_valtype_t* argTypes[argSize];
+
+    for (unsigned i = 0; i < argSize; ++i) {
+        argTypes[i] = wasm_valtype_new(func->funcArgTypes.at(i));
+    }
+
+    wasm_valtype_vec_t args;
+    wasm_valtype_vec_new(&args, argSize, argTypes);
+
+    size_t resultSize = func->funcResultTypes.size();
+    wasm_valtype_t* resultTypes[resultSize];
+
+    for (unsigned i = 0; i < resultSize; ++i) {
+        resultTypes[i] = wasm_valtype_new(func->funcResultTypes.at(i));
+    }
+
+    wasm_valtype_vec_t results;
+    wasm_valtype_vec_new(&results, resultSize, resultTypes);
+
+    own wasm_functype_t* externFuncType = wasm_functype_new(&args, &results);
+    own wasm_func_t* externFunc = wasm_func_new(_store, externFuncType, func->functionPointer);
+    wasm_extern_t* result = wasm_func_as_extern(externFunc);
```

Comment:
Why did you use `wasm-c-api` functions for wasi implementation?
IMO we can directly importing wasi functions without using any `wasm-c-api`
`wasm-c-api` is a list of c-api functions used for merging Walrus in other c-based project.

Reply:
I just thought it would be easier to handle wasi a bit more "outside" walrus, since it is in change right now and the api provides a good interface with that. My train of thought may be wrong though.

This whole implementation can be reworked to be entirely in Walrus.

Reply:
I'm not sure that using `wasm-c-api` is really efficiency.
WASI functions are imported to WASM modules like other imported functions.
so IMHO we can just define WASI functions, wrapping these functions as `ImportedFunction` and directly importing to WASM modules. That's all required for WASI implementation.

For now, its better to implement core functions of WASI such as defined in wabt
https://github.com/Samsung/walrus/blob/interp/third_party/wabt/include/wabt/interp/wasi_api.def

Reply:
You are right, it was a mistake on my part. I'll start to rework this patch to use `ImportedFunction`.